subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformationResult.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.transform;
+
+import com.google.common.collect.ImmutableList;
+
+import java.io.File;
+
+/**
+ * The result of running a transformation.
+ *
+ * The result of running a transformation is a list of outputs.
+ * There are two kinds of outputs for a transformation:
+ * - Produced outputs in the workspace. Those are absolute paths which do not change depending on the input artifact.
+ * - Selected parts of the input artifact. These are relative paths of locations selected in the input artifact.
+ */
+public interface TransformationResult {
+    /**
+     * Resolves location of the outputs of this results for a given input artifact.
+     *
+     * Produced outputs don't need to be resolved to locations, since they are absolute paths and can be returned as is.
+     * The relative paths of selected parts of the input artifact need to resolved based on the provided input artifact location.
+     */
+    ImmutableList<File> resolveOutputsForInputArtifact(File inputArtifact);
+
+    static Builder builder() {
+        return new Builder();
+    }
+
+    class Builder {
+        private final ImmutableList.Builder<TransformationOutput> builder = ImmutableList.builder();
+        boolean onlyProducedOutputs = true;
+
+        public void addInputArtifact(String relativePath) {
+            onlyProducedOutputs = false;
+            builder.add(new InInputArtifact(relativePath));
+        }
+
+        public void addInputArtifact() {
+            onlyProducedOutputs = false;
+            builder.add(new InputArtifact());
+        }
+
+        public void addOutput(File outputLocation) {
+            builder.add(new ProducedOutput(outputLocation));
+        }
+
+        public TransformationResult build() {
+            ImmutableList<TransformationOutput> transformationOutputs = builder.build();
+            return onlyProducedOutputs
+                ? new OnlyProducedOutputTransformationResult(convertToProducedOutputLocations(transformationOutputs))
+                : new DefaultTransformationResult(transformationOutputs);
```

Comment:
I think the difference here is that one is a `ResolvingTrnasformationResult` while the other is an `(Already)ResolvedTransformationResult`. What I don't like with the current names is the `Default`.

Reply:
I renamed the classes to `AlreadyResolvedTransformationResult` and `ResolvingTransformationResult`.